plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetInstallDelegate.java

```diff
@@ -10,32 +15,97 @@
-                                     },
-                                     true /* isExported */);
-      javaProject.setRawClasspath(newClasspath, monitor);
+      updateClasspath(project, monitor);
+    }
+
+  }
+
+  public static void installAppEngineRuntime(IFacetedProject project, IProgressMonitor monitor)
+      throws CoreException {
+    Set<IProjectFacetVersion> facets = new HashSet<>();
+    facets.add(WebFacetUtils.WEB_25);
+    Set<IRuntime> runtimes = RuntimeManager.getRuntimes(facets);
+    project.setTargetedRuntimes(runtimes, monitor);
+
+    if (RuntimeManager.isRuntimeDefined(AppEngineFacet.DEFAULT_RUNTIME_NAME)) {
+      IRuntime appEngineRuntime = RuntimeManager.getRuntime(AppEngineFacet.DEFAULT_RUNTIME_NAME);
+      project.setPrimaryRuntime(appEngineRuntime, monitor);
+    } else { // Create a new App Engine runtime
+      IRuntimeType appEngineRuntimeType = ServerCore.findRuntimeType(AppEngineFacet.DEFAULT_RUNTIME_ID);
+      if (appEngineRuntimeType == null) {
+        throw new NullPointerException("Could not find " + AppEngineFacet.DEFAULT_RUNTIME_NAME + " runtime type");
+      }
+
+      IRuntimeWorkingCopy appEngineRuntimeWorkingCopy
+          = appEngineRuntimeType.createRuntime(null, monitor);
+      File sdkLocation = new CloudSdkProvider(null).getCloudSdkLocation();
+      if (sdkLocation != null) {
+        IPath sdkPath = Path.fromOSString(sdkLocation.getAbsolutePath());
+        appEngineRuntimeWorkingCopy.setLocation(sdkPath);
+      }
+      org.eclipse.wst.server.core.IRuntime appEngineServerRuntime
+          = appEngineRuntimeWorkingCopy.save(true, monitor);
+      IRuntime appEngineFacetRuntime = FacetUtil.getRuntime(appEngineServerRuntime);
+      if (appEngineFacetRuntime == null) {
+        throw new NullPointerException("Could not locate App Engine facet runtime");
+      }
+
+      project.addTargetedRuntime(appEngineFacetRuntime, monitor);
+      project.setPrimaryRuntime(appEngineFacetRuntime, monitor);
     }
   }
 
+  public static void installAppEngineFacet(IFacetedProject facetedProject, IProgressMonitor monitor)
+      throws CoreException {
+    IFacetedProjectWorkingCopy workingCopy = facetedProject.createWorkingCopy();
+    IProjectFacet appEngineFacet = ProjectFacetsManager.getProjectFacet(AppEngineFacet.APP_ENGINE_FACET_ID);
+    IProjectFacetVersion appEngineFacetVersion = appEngineFacet.getVersion(AppEngineFacet.APP_ENGINE_FACET_VERSION);
+    workingCopy.addProjectFacet(appEngineFacetVersion);
+    workingCopy.commitChanges(monitor);
+  }
+
+  private void updateClasspath(IProject project, IProgressMonitor monitor) throws CoreException {
+    IJavaProject javaProject = JavaCore.create(project);
+    IClasspathEntry[] rawClasspath = javaProject.getRawClasspath();
+    IClasspathEntry[] newClasspath = new IClasspathEntry[rawClasspath.length + 1];
+    System.arraycopy(rawClasspath, 0, newClasspath, 0, rawClasspath.length);
+    newClasspath[newClasspath.length - 1] =
```

Comment:
It is adding a value to the last entry in newClasspath. If it was newClasspath[rawClasspath.length - 1] then it would be overwriting the second to  the last value in newClasspath